crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1166,18 +1179,138 @@ impl<'db> TypeInferenceBuilder<'db> {
+        fn inner<'db>(
+            builder: &mut TypeInferenceBuilder<'db>,
+            target: &ast::Expr,
+            value_ty: Type<'db>,
+            variable: &ast::ExprName,
+        ) -> Option<Type<'db>> {
+            match target {
+                // TODO: Handle other possible target expression like attribute, subscript,
+                // and starred.
+                ast::Expr::Name(name) if name == variable => {
+                    return Some(value_ty);
+                }
+                ast::Expr::List(ast::ExprList { elts, .. })
+                | ast::Expr::Tuple(ast::ExprTuple { elts, .. }) => match value_ty {
+                    Type::Tuple(tuple_ty) => {
+                        let starred_index = elts.iter().position(ast::Expr::is_starred_expr);
+
+                        let element_types = if let Some(starred_index) = starred_index {
+                            if tuple_ty.len(builder.db) >= elts.len() {
+                                let mut element_types = Vec::with_capacity(elts.len());
+                                element_types.extend_from_slice(
+                                    // SAFETY: Safe because of the length check above.
+                                    tuple_ty.elements(builder.db).get(..starred_index).unwrap(),
+                                );
+
+                                // E.g., in `(a, *b, c, d) = ...`, the index of starred element `b`
+                                // is 1 and the remaining elements after that are 2.
+                                let remaining = elts.len() - 2;
+                                let end_index =
+                                    tuple_ty.len(builder.db) - starred_index - remaining;
+                                // SAFETY: Safe because of the length check above.
+                                let _starred_element_types = tuple_ty
+                                    .elements(builder.db)
+                                    .get(starred_index..=end_index)
+                                    .unwrap();
+                                // TODO: Combine the types into a list type.
+                                // combine_types(starred_element_types);
+                                element_types.push(Type::Todo);
+
+                                element_types.extend_from_slice(
+                                    // SAFETY: Safe because of the length check above.
+                                    tuple_ty.elements(builder.db).get(end_index + 1..).unwrap(),
+                                );
+                                Cow::Owned(element_types)
+                            } else if tuple_ty.len(builder.db) == elts.len() - 1 {
+                                let mut element_types = tuple_ty.elements(builder.db).to_vec();
+                                element_types.insert(starred_index, Type::Todo);
+                                Cow::Owned(element_types)
```

Comment:
This is handling the case where the starred element would be bound to nothing, i.e. match zero elements and turn into an empty list? Maybe add a comment with an example of what it is handling.

I wonder if this case could just be integrated into the above case; we should be able to detect that `starred_index..=end_index` is an empty range and just handle that as needed (or just handle an empty `starred_element_types`.)

Reply:
> I wonder if this case could just be integrated into the above case; we should be able to detect that `starred_index..=end_index` is an empty range and just handle that as needed (or just handle an empty `starred_element_types`.)

Yeah, I think this is possible and should make it simpler. Thanks for the suggestion.